File: src/libraries/types/src/pool_stats.rs
use candid::{CandidType, Deserialize, Nat};
use serde::Serialize;

// Pool Snapshots

#[derive(CandidType, Deserialize, Clone, Serialize, Debug, PartialEq, Eq, Hash)]
pub struct PoolSnapshot {
    pub id: String,
    pub pool_id: String,
    pub timestamp: u64,
    pub position_data: Option<PositionData>,
    pub pool_data: Option<PoolData>,
}

#[derive(CandidType, Deserialize, Clone, Serialize, Debug, PartialEq, Eq, Hash)]
pub struct PositionData {
    pub id: Nat,
    pub amount0: Nat,
    pub amount1: Nat,
    pub usd_amount0: Nat,
    pub usd_amount1: Nat,
}

#[derive(CandidType, Deserialize, Clone, Serialize, Debug, PartialEq, Eq, Hash)]
pub struct PoolData {
    pub tvl: Nat,
    // pub balance0: Nat,
    // pub balance1: Nat,
    // pub lp_fee_0: Nat,
    // pub lp_fee_1: Nat,
}

// Pool Metrics

#[derive(CandidType, Deserialize, Clone, Serialize, Debug, PartialEq)]
pub struct PoolMetrics {
    pub apy: ApyValue,
    pub tvl: Nat,
}

#[derive(CandidType, Deserialize, Clone, Serialize, Debug, PartialEq)]
pub struct ApyValue {
    pub tokens_apy: f64,
    pub usd_apy: f64
}

File: frontend/src/idl/pool_stats.ts
import type { Principal } from '@dfinity/principal';
import type { ActorMethod } from '@dfinity/agent';
import type { IDL } from '@dfinity/candid';

export interface AddLiquidityResponse {
  'request_id' : bigint,
  'token_0_amount' : bigint,
  'token_1_amount' : bigint,
}
export interface ApyValue { 'tokens_apy' : bigint, 'usd_apy' : bigint }
export type ExchangeId = { 'Sonic' : null } |
  { 'KongSwap' : null } |
  { 'ICPSwap' : null };
export interface Pool {
  'id' : string,
  'provider' : ExchangeId,
  'initial_position' : [] | [Position],
  'token0' : Principal,
  'token1' : Principal,
}
export interface PoolApy {
  'month' : ApyValue,
  'week' : ApyValue,
  'year' : ApyValue,
}
export interface PoolByTokens {
  'provider' : ExchangeId,
  'token0' : Principal,
  'token1' : Principal,
}
export interface PoolData { 'tvl' : bigint }
export interface PoolMetrics { 'apy' : PoolApy, 'tvl' : bigint }
export interface PoolSnapshot {
  'id' : string,
  'pool_data' : [] | [PoolData],
  'timestamp' : bigint,
  'pool_id' : string,
  'position_data' : [] | [PositionData],
}
export interface PoolSnapshotArgs {
  'pool_data' : [] | [PoolData],
  'timestamp' : bigint,
  'pool_id' : string,
  'position_data' : [] | [PositionData],
}
export interface Position {
  'id' : bigint,
  'initial_amount0' : bigint,
  'initial_amount1' : bigint,
}
export interface PositionData {
  'id' : bigint,
  'usd_amount0' : bigint,
  'usd_amount1' : bigint,
  'amount0' : bigint,
  'amount1' : bigint,
}
export type Result = { 'Ok' : AddLiquidityResponse } |
  { 'Err' : string };
export type Result_1 = { 'Ok' : WithdrawFromPoolResponse } |
  { 'Err' : string };
export interface WithdrawFromPoolResponse {
  'token_0_amount' : bigint,
  'token_1_amount' : bigint,
}
export interface _SERVICE {
  'add_liquidity_to_pool' : ActorMethod<[string, bigint], Result>,
  'add_pool' : ActorMethod<[Principal, Principal, ExchangeId], string>,
  'add_pool_snapshot' : ActorMethod<[PoolSnapshotArgs], undefined>,
  'delete_all_pools_and_snapshots' : ActorMethod<[], boolean>,
  'delete_pool' : ActorMethod<[string], boolean>,
  'delete_pool_snapshot' : ActorMethod<[string, string], undefined>,
  'delete_pool_snapshots' : ActorMethod<[string], undefined>,
  'get_pool_by_id' : ActorMethod<[string], [] | [Pool]>,
  'get_pool_by_tokens' : ActorMethod<[PoolByTokens], [] | [Pool]>,
  'get_pool_metrics' : ActorMethod<
    [Array<string>],
    Array<[string, PoolMetrics]>
  >,
  'get_pools' : ActorMethod<[], Array<Pool>>,
  'get_pools_snapshots' : ActorMethod<
    [Array<string>],
    Array<[string, Array<PoolSnapshot>]>
  >,
  'remove_liquidity_from_pool' : ActorMethod<[string], Result_1>,
  'set_operator' : ActorMethod<[Principal], undefined>,
  'update_pool_ids' : ActorMethod<[], boolean>,
}
export declare const idlFactory: IDL.InterfaceFactory;
export declare const init: (args: { IDL: typeof IDL }) => IDL.Type[];
